### DISCUSSION TITLE
現行の翻訳のkeyの取得方法と、今後の改善点

### DISCUSSION THREAD

**Discussion by L4Ph** [Katsuyuki-Karasawa/stable-diffusion-webui-localization-ja_JP#26] [General]
#23 から
いまは

1. 翻訳用のWebUI入れる
2. `git pull`→
3. まっさらな環境の翻訳のkeyを出力→
4. 欲しい拡張機能の翻訳やアップデートの差分を取得→
5. 再度keyを出力→
6. リセット

という状態
そのため、下手にテンプレートになるkeyをいじれないのがネック

**Comment by L4Ph** [comment]
- [x] 翻訳済みの拡張機能のチェックボックス(リスト)の設置
- [x] 翻訳の差分の取得
- [x] crowdin以外も視野に入れるべきか
- [x] 他の言語をベースに+バイリンガルへの切替(別ブランチ)

**Reply by L4Ph** [reply]
バイリンガルをそもそもこちらに取り込んでもいい気がしてきた
が、submoduleで特定のファイルを取得するのは微妙に厄介

**Reply by L4Ph** [reply]
wgetとかcurl+actionsで取得するのが早そう?

**Reply by L4Ph** [reply]
チェックボックスやりたいけど、微妙にリスト作るのが面倒
ファイル名から適当にバッチ処理書いて取得する(そのうち)

**Reply by L4Ph** [reply]
harukaxxxx氏のおかげでチェックリストの設置ができた
Thanks! @harukaxxxx

**Comment by L4Ph** [comment]
#25 これをactionsで回せるようにするだけでだいぶ楽そうなのでそれをやる

**Reply by L4Ph** [reply]
[zh-TW](https://github.com/benlisquare/stable-diffusion-webui-localization-zh_TW)の手法でいい気もしてきた

**Reply by L4Ph** [reply]
実装した
[source](https://github.com/benlisquare/stable-diffusion-webui-localization-zh_TW/tree/main/json/source)からとってくる
あとはマージするように変更します
cronでやるので1日1回回るはず

**Reply by L4Ph** [reply]
cron入れた
0時と12時の2回
これでソースの自動更新は行われるはず

**Reply by Rokurolize** [reply]
返信が遅くなり申し訳ありません！不明点を解決してからコメントする予定でしたが、時間がかかりそうなので現時点で理解できている範囲でコメントします。

以下のトピックについては、まだ完全に理解できていないため、コメントが不十分な部分があるかもしれません。

- Crowdinの機能
- GitHub Actionsの機能
- さきほど唐澤さんが言及されていたzh-TWの手法

#25 に関しては、機能（JSONのソートとマージ）だけを用意し、運用方法は後で詰める予定でした。その後、唐澤さんのコメントでGitHub Actionsの存在を知り、非常に便利だと感じました。

新しいキーの追加に関して、現時点で把握している情報を以下に記載します。

- WebUIで出力した翻訳テンプレートファイルは、WebUI起動後に読み込まれた順に文字列が蓄積されています（WebUIのコードは追跡できていないため、正確な挙動は不明です）。
    - 環境やWebUIのバージョンによって並び順が変わる可能性があるため、リポジトリに置くJSONは常にソートしておくことが望ましいです。
- WebUIで一度も表示されていない文字列は、出力時に翻訳テンプレートに含まれません。
    - 以下のような文字列が該当します。
        - 拡張機能の更新チェックで更新が見つかると表示される「behind」「latest」などの文字列
ただし、WebUIで一度表示されれば、その後の出力で翻訳テンプレートに含まれるようになります。
        - ローカライズ用ファイルには含まれているが、WebUIで未インストールの拡張機能に含まれる文字列

これらを考慮すると、以下の運用方法が適切と考えられます。

### ディレクトリ構造
```
repo/
├── localizations/
│   └── ja_JP.json
└── template/
    ├── template.json
    └── extensions/
        ├── extension1.json
        ├── extension2.json
        └── ... (その他の拡張機能のjsonファイル)
```

### 処理内容
1. `repo/localizations/ja_JP.json`をベースに、 `repo/template/extensions`内のすべてのJSONファイルを`merge_json_files.py`でマージし、`repo/template/template.json`に出力します。
1. `repo/template/template.json`をCrowdinで読み込み、翻訳作業を行います。
1. 翻訳作業が終わったら、Crowdinが`ja_JP.json`を`repo/localizations/ja_JP.json`に上書きし、リポジトリを更新します。

上記で述べた「behind」「latest」のように、WebUIで出力したテンプレートファイルから漏れる文字列が存在する場合があります。例えば、`repo/template/extensions/extension1.json`に関して未翻訳の文字列が見つかった場合は、GitHubでプルリクエストを行い、手動でキーを追加することを想定しています。その際も、ソートを行ってからコミットすることで、差分の表示が綺麗になると考えられます。

これからzh-TWの手法について確認してみます。何かフィードバックがあればぜひお知らせください。

**Reply by harukaxxxx** [reply]
> ### 処理内容
> 1. `repo/localizations/ja_JP.json`をベースに、 `repo/template/extensions`内のすべてのJSONファイルを`merge_json_files.py`でマージし、`repo/template/template.json`に出力します。
> 2. `repo/template/template.json`をCrowdinで読み込み、翻訳作業を行います。
> 3. 翻訳作業が終わったら、Crowdinが`ja_JP.json`を`repo/localizations/ja_JP.json`に上書きし、リポジトリを更新します。

処理内容の 1. は、Crowdin で昔翻訳済みの json 各拡張機能にアップロードすると、自動的にマージしますよ、後は目を通してだけです。
Crowdin の Translations タップに Upload existing translations の機能を使ている、ja_JP.json をアップロードの後 Source file は指定の拡張機能に選択、Target language も選択して import！

**Reply by Rokurolize** [reply]
考え中のこと

Crowdinの翻訳画面で拡張機能ごとに項目が分かれているのは便利だと思うが、その翻訳結果がリポジトリのextensionsフォルダ配下のファイルにまで反映されていくのは煩わしく感じる。翻訳結果はja_JP.jsonだけあれば良いはず。また、キーの追加を行うコミットと翻訳を反映するコミットがぐちゃぐちゃになるので、この2種類を分ける意味合いでリポジトリが別々になればすっきりしそうな気がする。

#47 のような問題を避けたいので、別リポジトリからキー項目を取ってきているなら、このリポジトリの中に別リポジトリを内包して参照した方が良いように思える（どうやるのかは分からないが、 https://github.com/derrian-distro/LoRA_Easy_Training_Scripts では実現しているように見える）。キーの追加はその別リポジトリに対してPRを行えば二度手間にならない。

**Reply by harukaxxxx** [reply]
I also noticed that commit will be too long in history, I have check into [crowdin github integration](https://support.crowdin.com/github-integration/) page, in QA section I find a recommendations for this problem
> "Squash commits when merging a localization branch to keep the master branch history clean and uncluttered."
![image](https://user-images.githubusercontent.com/5777085/226635257-ae5ee052-51db-4bef-affd-dd005db681eb.png)

Maybe it will solve this problem by using squash commit? I have no idea what will it be since I haven't do this before.

**Reply by L4Ph** [reply]
@Rokurolize Submoduleの使用は考えましたが、少なくとも現段階のCrowdinの仕様として使うことができません
また、今回の問題点はファイル名が別であることから発生しているので、個人的にはどうしようもないと思います(そもそも別個のファイルとして扱われるので)

**Reply by harukaxxxx** [reply]
> I also noticed that commit will be too long in history, I have check into [crowdin github integration](https://support.crowdin.com/github-integration/) page, in QA section I find a recommendations for this problem
> 
> > "Squash commits when merging a localization branch to keep the master branch history clean and uncluttered."
> > ![image](https://user-images.githubusercontent.com/5777085/226635257-ae5ee052-51db-4bef-affd-dd005db681eb.png)
> 
> Maybe it will solve this problem by using squash commit? I have no idea what will it be since I haven't do this before.

Squash commits [works pretty fine](https://github.com/harukaxxxx/stable-diffusion-webui-localization-zh_TW/commit/86ac2f6fc6738497174de0b13155e31de2b171ad) !! @Katsuyuki-Karasawa should do this when approval PR

![image](https://user-images.githubusercontent.com/5777085/226669018-8b17aed6-8ad1-47a2-8f66-e3e2f7fcfd7c.png)

**Reply by Rokurolize** [reply]
> Submoduleの使用は考えましたが、少なくとも現段階のCrowdinの仕様として使うことができません

そうなんですね…了解です。



> 今回の問題点はファイル名が別であることから発生しているので、個人的にはどうしようもないと思います

curlでファイルを取ってきているからgitのように名前変更やファイルの削除があった場合に追跡できない問題があるという認識だったのですが、これはGitHub Actionsの中でgitで取ってきて管理したりできないのでしょうか。

**Reply by harukaxxxx** [reply]
確かにその辺は全て自動化に任せるにはいけませんですね、少し人工的の作業は必要と思います。
ソースをダウンロードする際に元のソースファイルと翻訳済みのファイルを一旦除去し、crowdinから全部再同調が解決できるが（crowdin_mainも除去するかも）、怖いのであまり頻繁にすることはない。
ソースのメインテーナスはファイル名を変更しないように気をつける🫠

**Reply by L4Ph** [reply]
这有点暴力，但在运行`rm -rf /template/source/`后，理论上你可以替换所有的东西 🤔 @harukaxxxx 
少し暴力的ですが、`update-source.yaml`に`rm -rf /template/source/`を仕込むといいのかもしれない

**Reply by L4Ph** [reply]
- 同じファイル名→`rm -rf /template/source`をしても勝手にリモートでは追跡される
- ファイル名が変わったもの→`/template/source`に新規に追跡されるが、ファイル名がまるきり変わっているのでそれまでのファイル名変更前の翻訳が消えてしまう
- いらないファイル→勝手に消える

こんなところでしょうか問題は

**Reply by L4Ph** [reply]
2番目が厄介?
まぁでもこれはpushする前に置き換えればいいのかもしれない

もしくは、mainとは別のブランチ作っておいてそっちで稼働させるか

**Reply by harukaxxxx** [reply]
#52 主な問題は、ソースファイルは削除および更新されているが、crowdinは存在しないソースファイルに対応する翻訳ファイルを削除しないため、Unpromptとunpromptの大文字と小文字が異なることが原因でgitエラーが発生することです。本当の原因は、リネーム後に古い翻訳ファイルが削除されていないことです。マージ手順ではこの問題を無視できますが、知らずにja_JP.jsonに間違った翻訳をマージする可能性があります。

過去、この問題を解決するために、crowdinのブランチを含むすべての翻訳ファイルを削除し、新しい翻訳をcrowdinから同期して完全な翻訳ファイルを再構築する必要がありました。しかし、実際には、どのファイルが不要かをすでに知っている場合は、理論的にはja_JPの下で手動でそれらを削除するだけで十分です（mainおよびcrowdin_mainの両方）。

**Reply by L4Ph** [reply]
@harukaxxxx done
https://github.com/Katsuyuki-Karasawa/stable-diffusion-webui-localization-ja_JP/commit/5bed9487b2301806be29b109acfad7e0bcbfd428

**Comment by L4Ph** [comment]
とりあえず100%到達したので一安心

**Reply by L4Ph** [reply]
次にやる作業を、思い出し次第羅列する
